c10/core/Allocator.h

```diff
@@ -230,6 +231,19 @@ struct C10_API InefficientStdFunctionContext {
   std::function<void(void*)> deleter_;
   InefficientStdFunctionContext(void* ptr, std::function<void(void*)> deleter)
       : ptr_(ptr), deleter_(std::move(deleter)) {}
+  InefficientStdFunctionContext(const InefficientStdFunctionContext&) = delete;
+  InefficientStdFunctionContext(InefficientStdFunctionContext&& rhs) noexcept
+      : ptr_(rhs.ptr_), deleter_(std::move(rhs.deleter_)) {
+    rhs.ptr_ = nullptr;
+  }
+  InefficientStdFunctionContext& operator=(
+      const InefficientStdFunctionContext&) = delete;
+  // NOLINTNEXTLINE(performance-noexcept-move-constructor)
+  InefficientStdFunctionContext& operator=(
```

Comment:
Why shouldn't this be noexcept?

Reply:
Because the destructor should be called first and there is no way to ensue that it is noexcept. The deleter may throw.

Reply:
I'm still a bit confused, whether the destructor is called or not is orthogonal to whether this method is noexcept? This is just calling std::move which doesn't necessarily call the destructor on RHS, the destroy should be called outside of the context of this assignment, correct?